text/3617-precise-capturing.md

```diff
@@ -0,0 +1,668 @@
+//~^ ERROR cannot return value referencing function parameter `x`
+}
+```
+
+However, if we know we have a value of type `B`, we can *rely* on the fact that the lifetime was not used:
+
+```rust
+fn test_b(x: B) -> impl Sized + 'static {
+    x.foo() //~ OK.
+}
+```
+
+We would say that the impl for `B` is *refining* in that it offers more to or demands less of callers than the minimum the trait could offer or the maximum it could demand.  Associated type definitions are always refining in this way.
+
+RPITIT desugars into associated types similar to those above, but here we've currently decided to lint against this refinement, e.g.:
+
+```rust
+trait Trait {
+    fn foo(&self) -> impl Sized;
+}
+
+impl Trait for () {
+    fn foo(&self) -> () {}
+//~^ WARN impl trait in impl method signature does not match
+//~|      trait method signature
+//~| NOTE add `#[allow(refining_impl_trait)]` if it is intended
+//~|      for this to be part of the public API of this crate
+//~| NOTE we are soliciting feedback, see issue #121718
+//~|      <https://github.com/rust-lang/rust/issues/121718>
+//~|      for more information
+}
+```
+
+Similarly, for consistency, we'll lint against RPITIT cases where less is captured by RPIT in the impl as compared with the trait definition when using `use<..>`.  E.g.:
+
+```rust
+trait Trait {
+    fn foo(&self) -> impl Sized;
+}
+
+impl Trait for () {
+    fn foo(&self) -> impl use<> Sized {}
+//~^ WARN impl trait in impl method signature does not match
+//~|      trait method signature
+//~| NOTE add `#[allow(refining_impl_trait)]` if it is intended
+//~|      for this to be part of the public API of this crate
+//~| NOTE we are soliciting feedback, see issue #121718
+//~|      <https://github.com/rust-lang/rust/issues/121718>
+//~|      for more information
+}
+```
+
+## Argument position impl Trait
+
+Note that for a generic type parameter to be captured with `use<..>` it must have a name.  Anonymous generic type parameters introduced with argument position `impl Trait` (APIT) syntax don't have names, and so cannot be captured with `use<..>`.  E.g.:
+
+```rust
+fn foo(x: impl Sized) -> impl use<> Sized { x }
+//                       ^^^^^^^^^^^^^^^^
+//                       ^ Captures nothing.
+```
+
+## Migration strategy for Lifetime Capture Rules 2024
+
+The migration lints for Rust 2024 will insert `use<..>` as needed so as to preserve the set of generic parameters captured by each RPIT opaque type.  That is, we will convert, e.g., this:
+
+```rust
+//@ edition: 2021
+fn foo<'t, T>(_: &'t (), x: T) -> impl Sized { x }
+```
+
+...into this:
+
+```rust
+//@ edition: 2024
+fn foo<'t, T>(_: &'t (), x: T) -> impl use<T> Sized { x }
+```
+
+Note that since generic type parameters must have names to be captured with `use<..>`, some uses of APIT will need to be converted to named generic parameters.  E.g., we will convert this:
+
+```rust
+//@ edition: 2021
+fn foo<'t>(_: &'t (), x: impl Sized) -> impl use<T> Sized { x }
```

Comment:
```suggestion
fn foo<'t>(_: &'t (), x: impl Sized) -> impl Sized { x }
```

there is no `T` and I suppose this is to demonstrate the insertion of `T: Sized` after the edition upgrade `cargo fix`.

Reply:
Thanks, good catch, fixed.